main/src/main/scala/sbt/Defaults.scala

```diff
@@ -154,6 +161,8 @@ object Defaults extends BuildCommon {
       scalaHome :== None,
       apiURL := None,
       javaHome :== None,
+      discoveredJavaHomes := discoverJavaHomes.value ++ additionalJavaHomes.value,
```

Comment:
This definition isn't intuitive to me. Discovered should only be discovered.

Maybe there should be a `fullJavaHomes` (like `fullResolvers` and `fullClasspath`) with this RHS.

Reply:
Sounds good. Then `fullJavaHomes = discoveredJavaHomes + javaHomes`.

Reply:
Because `javaHomes` is not clear if it's the unmanaged part of the full.

```scala
javaHomes := Map("10" -> something)
```

I am ok with either `discoveredJavaHomes` or `managedJavaHomes` since the build user won't be interacting with them that much.

Reply:
> Because `javaHomes` is not clear if it's the unmanaged part of the full.

It's the shorter, user-facing one, like I believe `resolvers` was designed to be.

Reply:
With `resolvers` the build users are somewhat aware that sbt will hit Maven Central by default and that you are supplying additional not-so-common custom resolvers for your subproject. In the case of `javaHomes`, if you _had_ to change it, that means sbt failed to detect some JDK homes automatically, and you are augmenting it with your machine-dependent info. IMO it should actually be some longer "it only works on your machine" sounding name. If not `unmanagedJavaHomes` it might be `myLittleJavaHomes`.

Reply:
> if you _had_ to change it, that means sbt failed to detect some JDK homes automatically, and you are augmenting it with your machine-dependent info.

I see it as the same as `resolvers`. You're adding not-so-common custom java homes paths.

Also, like `resolvers`, it could have "doesn't work for everyone" concerns (e.g company-specific artefact repository URLs and company-specific JVM installation paths).

Reply:
> `resolvers` typically go into build.sbt whereas your `Global / (unmanaged)JavaHomes` should go into user-defined global settings; never in `build.sbt`.

I can see both. For instance if you're in a company that uses a custom JDK/JVM installed (e.g at Twitter with GraalVM) in a specific but not-so-common path I can see it being experimented with at a project level, maybe eventually being defined in a company sbt plugin. Not necessarily in user-wide global settings.

And even if it were meant for user-wide global settings, why shouldn't it still be called `javaHomes`?

> whereas if JDK cross building doesn't work, the users will need to type `show Global/fullJavaHomes` to debug why it wasn't detected.

To debug what was discovered they would look at `discoveredJavaHomes`.

> If we really want to use the name `javaHomes`, I would say we should get rid of `fullJavaHomes`, and just use `javaHomes` key where the user can just `+=` into it.

I don't follow, but I think it's non sequitur from my earlier comments.

Reply:
<img width="1098" alt="things_and_cothings" src="https://user-images.githubusercontent.com/184683/39479502-72108a7e-4d33-11e8-9712-8a2c24af1ddf.png">

Reply:
We discussed in sbt 1 meeting today. I'll take this PR and make the edits.